### DISCUSSION TITLE
What is a Job?

### DISCUSSION THREAD

**Discussion by paulius-hydrosat** [dagster-io/dagster#16306] [Q&A]
I'm working on https://github.com/dagster-io/dagster/issues/16129 and I'm stumped by the `get_base_asset_jobs` logic. I would appreciate any hints on what kind of object is a Job supposed to be in the context of software-defined assets. To make this a bit more concrete, here are some aspects that I find puzzling.

1. Consider this graph:
![What is Job ](https://github.com/dagster-io/dagster/assets/135983426/b45c6600-b4df-450e-8c08-70f864f54da0)

For it, 3 "base asset" jobs are created. The logic for adding nodes to these jobs are as follows - groupby partition_def, add all unpartitioned. So ex. `__ASSET_JOB_2` contains asset_xyz_1 and asset_xyz_2, even though they are unrelated in the DAG sense. What purpose could such jobs have? My "normal" materialization pattern would be to shift click asset_0, asset_abc and asset_123 and materialize them. But they don't belong to a single job?

2. There is a function `define_asset_job`, what would I use it for? If I create a job using that function, all the other "base asset" jobs still get created. What would happen if I created tons of arbitrary jobs? Nothing?

3. Jobs doc page starts with a statement "Jobs are the main unit of execution and monitoring in Dagster". Is that true for software defined assets? Isn't "Run" the main unit of execution and monitoring? Or "Step"?

4. Now consider this graph:
![image](https://github.com/dagster-io/dagster/assets/135983426/c369f6a6-86bc-47c4-8d43-7cdb396096fd)

It is a simple one, with no partitioned asset 0 and partitioned asset abc. By default, if I shift click both of them and click materialize I get 4 runs: asset_0, asset_abc partition a, partition b, partition c. Now I hack the code a little: I modify the "get_base_asset_jobs" function to return empty list, and create a job myself using `jobs = [define_asset_job('job', [asset_0, asset_abc])]` and pass jobs to Definitions. Now when I shift click them both and click materialize I get 3 runs: every run materializes asset_0 and on of the partitions of abc. Why? Isn't it the same job that I created?

**Comment by sryza** [comment]
Hi @paulius-hydrosat - from the perspective of a users, a job is an object created using `define_asset_job`, the `@job` decorator, or similar.  I.e. it's an explicitly-named DAG of ops or assets that shows up in the UI and can be executed atomically.

From the perspective of Dagster internals, not meant to be encountered by most users, a job is something that can "be run". So we use the hidden asset jobs that you're noticing to hold groups of assets that it's possible to launch runs for.

Dagster currently has a constraint that each run targets a single partition. This means that, if different assets have different `PartitionsDefinition`s, they can't be in the same run, and thus can't be in the same job. This is why we create a hidden asset job for each `PartitionsDefinition`

A couple changes that we could imagine making in the future:
- Enable a single run to target multiple partitions. This would allow us to consolidate on a single hidden asset job per code location.
- Enable runs of assets without any job involved. This would allow us to remove the hidden asset jobs.

**Reply by psarka** [reply]
Thanks!
